pkg/controller/endpoint/trigger_time_tracker.go

```diff
@@ -0,0 +1,147 @@
+	"sync"
+	"time"
+
+	"k8s.io/api/core/v1"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+)
+
+// TriggerTimeTracker is a util used to compute the EndpointsLastChangeTriggerTime annotation which
+// is exported in the endpoints controller's sync function.
+// See the documentation of the EndpointsLastChangeTriggerTime annotation for more details.
+//
+// Please note that this util may compute a wrong EndpointsLastChangeTriggerTime if a same object
+// changes multiple times between two consecutive syncs. We're aware of this limitation but we
+// decided to accept it, as fixing it would require a major rewrite of the endpoints controller and
+// Informer framework. Such situations, i.e. frequent updates of the same object in a single sync
+// period, should be relatively rare and therefore this util should provide a good approximation of
+// the EndpointsLastChangeTriggerTime.
+// TODO(mm4tt): Implement a more robust mechanism that is not subject to the above limitations.
+type TriggerTimeTracker struct {
+	// endpointsStates is a map, indexed by namespaced Endpoints name storing the last known
+	// Endpoints object state observed during the most recent call of the
+	// ComputeEndpointsLastChangeTriggerTime function.
+	endpointsStates map[string]endpointsState
+
+	// mutex guarding the endpointsStates map.
+	mutex sync.Mutex
+}
+
+// NewTriggerTimeTracker creates a new instance of the TriggerTimeTracker.
+func NewTriggerTimeTracker() *TriggerTimeTracker {
+	return &TriggerTimeTracker{
+		endpointsStates: make(map[string]endpointsState),
+	}
+}
+
+// endpointsState represents a state of an Endpoints object that is known to this util.
+type endpointsState struct {
+	// lastServiceTriggerTime is a service trigger time observed most recently.
+	lastServiceTriggerTime time.Time
+	// lastPodTriggerTimes is a map (Pod name -> time) storing the pod trigger times that were
+	// observed during the most recent call of the ComputeEndpointsLastChangeTriggerTime function.
+	lastPodTriggerTimes map[string]time.Time
+}
+
+// ComputeEndpointsLastChangeTriggerTime updates the state of the Endpoints object being synced
+// and returns the time that should be exported as the EndpointsLastChangeTriggerTime annotation.
+//
+// Important: This method is go-routing safe but only when called for different keys. The method
+// shouldn't be called concurrently for the same key! This contract is fulfilled in the current
+// implementation of the endpoints controller.
+func (t *TriggerTimeTracker) ComputeEndpointsLastChangeTriggerTime(
+	key string, service *v1.Service, pods []*v1.Pod) time.Time {
+
+	// As there won't be any concurrent calls for the same key, we need to guard access only to the
+	// endpointsStates map.
+	t.mutex.Lock()
+	state, wasKnown := t.endpointsStates[key]
+	t.mutex.Unlock()
+
+	// Update the state before returning.
+	defer func() {
+		t.mutex.Lock()
+		t.endpointsStates[key] = state
+		t.mutex.Unlock()
+	}()
+
+	var minChangedTriggerTime time.Time
```

Comment:
Probably add a piece of comment here to explain why using the minChangedTriggerTime to do the measurement may not yeild the intended result due to multiple update getting merged into one. 

Basically,    sync1 ->  pod1, pod2, pod3 changed -> sync2    
`minChangedTriggerTime` in this case is time delta between pod3 update and sync2, but the actual time difference may be delta between pod1 update and sync2.

Reply:
You're right, assuming that by pod1, pod2, pod3 you're referring to multiple updates of the same pod. 
In such case here we'd see and export the pod3 change time whereas we should've used the pod1 change time.
This was already documented in the TriggerTimeTracker struct documentation, but added it also in the doc of this function.

And just to make one thing clear, we're not operating on time deltas here. Here we only export the "trigger time" which is the time of the oldest change that triggered endpoints object change. Time delta will be computed in kube-proxy to export the actual latency.